File: C++/binary search.cpp
#include <iostream>
#include <vector>
using namespace std;

// Binary Search assumes array is sorted
int binarySearch(const vector<int>& nums, int target) {
    int left = 0, right = nums.size() - 1;

    while (left <= right) {
        int mid = left + (right - left) / 2;  // Prevent overflow

        if (nums[mid] == target)
            return mid;  // Target found
        else if (nums[mid] < target)
            left = mid + 1;  // Search right half
        else
            right = mid - 1; // Search left half
    }
    return -1;  // Not found
}

// ✅ Use-case
int main() {
    vector<int> nums = {1, 3, 5, 7, 9};
    int target = 5;

    int result = binarySearch(nums, target);
    cout << "Index of 5: " << result << endl;  // Output: 2
}

File: Java/binary search.java
public class BinarySearch {

    // Returns index if found, else -1
    public static int binarySearch(int[] nums, int target) {
        int left = 0, right = nums.length - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2;  // Prevent overflow

            if (nums[mid] == target)
                return mid;
            else if (nums[mid] < target)
                left = mid + 1;  // Search right
            else
                right = mid - 1; // Search left
        }

        return -1;  // Not found
    }

    public static void main(String[] args) {
        int[] nums = {2, 4, 6, 8, 10, 12};  // Must be sorted
        int target = 8;
        int index = binarySearch(nums, target);

        System.out.println("Index of " + target + ": " + index);  // Output: 3
    }
}
